### PR TITLE
Fix 5780

### LINKED ISSUE #5780: add creating, editing and displaying placements to casa case show page
## What type(s) of user does this feature affect?
 - supervisors
 - admins
 - volunteers

## Description

Part of #4695

We currently have a partially implemented Placements feature. Placements are essentially a combination way to track with whom and where a case subject is living and for how long.

This ticket is a bit larger and more open ended than some of the related ones. We want to be able to do all the usual CRUD stuff to placements. 

I will leave it up to you as to how best to display all that functionality, the only requirement is to have that on the case page (`/casa_cases/:id`)

### Some Ideas (PLEASE don't actually hard copy my mockups they are hideous 🤮)

Model them after court orders: have all the adding, editing and deleting on the case page (volunteers should be able to set these)

Image

Create an index page that shows all the placements for a case. Create edit and show pages for specific placements. But only show the current placement on the Case page. It can then have a kebab menu for other actions.

Image

## How to access the QA site
_Login Details:_  
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin
- all_casa_admin1@example.com view site as an all casa admin
  - go to `/all_casa_admins/sign_in`  

password for all users: 12345678  

### Questions? Join Slack!

We highly recommend that you join us in [slack](https://join.slack.com/t/rubyforgood/shared_invite/zt-21pyz2ab8-H6JgQfGGI0Ab6MfNOZRIQA) #casa channel to ask questions quickly. And [discord](https://discord.gg/qJcw2RZH8Q) for office hours (currently Tuesday 5-7pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by guswhitten**
### What github issue is this PR for, if any?
Resolves #5780

### What changed, and _why_?
Add basic CRUD operations for case Placements.

### How is this **tested**? (please write tests!) 💖💪
- `index.html.erb_spec.rb`
- `edit.html.erb_spec.rb`
- `new.html.erb_spec.rb`

### Screenshots please :)
placement-partial-updated
index-view
edit-view
delete-modal

### CODE REVIEW

File: app/models/casa_org.rb
Comment (guswhitten) [inline app/models/casa_org.rb:31]: unless i'm mistaken, there's currently no way for placement types to be created for an org. should i add that work to this PR or should another issue be made?

File: app/views/casa_cases/_placements.html.erb
Comment (guswhitten) [inline app/views/casa_cases/_placements.html.erb:1]: i thought the look of the table would be clean, but i am not all that satisfied with the look.  i'm open to suggestions

File: app/models/casa_org.rb
Comment (elasticspoon) [reply app/models/casa_org.rb:31]: Case placements are part of a larger epic https://github.com/rubyforgood/casa/issues/4695. https://github.com/rubyforgood/casa/issues/5778 that issue deals with the creation of placements at an org level

File: app/views/casa_cases/_placements.html.erb
Comment (elasticspoon) [inline app/views/casa_cases/_placements.html.erb:1]: I think we want the case page to be a bit simpler. 

Instead of table how about just 

Current Placement: placement_type.name
Placed Since: placement_started_at
See All => link to index

and if a user wants to interact with placements they can just do all the actions on the index page.

File: app/assets/stylesheets/pages/casa_cases.scss
Comment (elasticspoon) [inline app/assets/stylesheets/pages/casa_cases.scss:1]: can this be done with inline bootstrap styles?

**Review by elasticspoon — CHANGES_REQUESTED** [review]
Hey. This is a great start!

Overall, my comments are more about removal of code than anything. 

I also would like to see a system spec to exercise the major work flows => creating, editing and deleting a placement.

File: app/controllers/placements_controller.rb
Comment (thejonroberts) [inline app/controllers/placements_controller.rb:8]: Typically, no `authorize` call in index action. Better to just policy_scope the collection of records.  You will need to implement a PlacementPolicy::Scope, and then:

```suggestion
    @placements = policy_scope(@casa_case.placements).includes(:placement_type).order(placement_started_at: :desc)
```

File: app/controllers/placements_controller.rb
Comment (thejonroberts) [inline app/controllers/placements_controller.rb:17]:
```suggestion
    @placement = Placement.new(casa_case: @casa_case)
    authorize @placement
```

File: app/models/placement_type.rb
Comment (thejonroberts) [inline app/models/placement_type.rb:5]: I'd prefer separate scopes for ordering, especially if it's not in the scope name!

**Review by thejonroberts — COMMENTED** [review]
I have to run before finished reviewing, have some minor suggestions.

File: app/controllers/placements_controller.rb
Comment (thejonroberts) [reply app/controllers/placements_controller.rb:8]: You could do both, like if limiting the view to admin/supervisor, with `authorize :placement` or `authorize Placement`

File: config/environments/test.rb
Comment (thejonroberts) [inline config/environments/test.rb:12]:
```suggestion
```

File: config/environments/test.rb
Comment (thejonroberts) [inline config/environments/test.rb:21]:
```suggestion

# cache classes on CI, but enable reloading for local work (bin/rspec)
  config.enable_reloading = ENV["CI"].blank?
```

I mad this change recently, apparently missed the other setting above while doing so. Can you  revert the changes in this file or put this back in? What was the issue here?

File: spec/factories/casa_orgs.rb
Comment (thejonroberts) [inline spec/factories/casa_orgs.rb:17]:
```suggestion
      transient { placement_names { ["Reunification", "Adoption", "Foster Care", "Kinship"] } }
```
Confusing, these aren't placement types, they are just type names.

File: config/environments/test.rb
Comment (guswhitten) [reply config/environments/test.rb:21]: Did not mean to commit this. I enable reloading locally so i can run tests.

File: spec/factories/casa_orgs.rb
Comment (thejonroberts) [inline spec/factories/casa_orgs.rb:23]:
```suggestion
      placements do
        placement_names.map { |name| association(:placement_type, name:) }
      end
```

I _**think**_ this would work, and play well with `build`, `build_stubbed`, or `create`. [see here](https://thoughtbot.github.io/factory_bot/cookbook/has_many-associations.html)

Not a huge deal, but I do think the placement_names change should be made at the least:

```suggestion
      after(:create) do |org, evaluator|
        evaluator.placement_names.each do |name|
          org.placement_types.create!(name: name)
        end
      end
```

File: spec/system/placements/destroy_spec.rb
Comment (thejonroberts) [inline spec/system/placements/destroy_spec.rb:28]: for same reason:
```suggestion
    expect { click_on "Confirm" }.to change(Placement, :count).by(-1)
```

File: spec/system/placements/edit_spec.rb
Comment (thejonroberts) [inline spec/system/placements/edit_spec.rb:17]:
```suggestion
    visit edit_casa_case_placement_path(casa_case, placement)
```

I think would work - can run `bin/rails routes -g casa_case_placement` to check it's there. no big deal, just on less page load for every test.

File: spec/policies/placement_policy_spec.rb
Comment (thejonroberts) [inline spec/policies/placement_policy_spec.rb:1]: 👍🏻

File: spec/system/placements/new_spec.rb
Comment (thejonroberts) [inline spec/system/placements/new_spec.rb:17]:
```suggestion
    visit new_casa_case_placements_path(casa_case)
```
should work here.

File: spec/views/notifications/index.html.erb_spec.rb
Comment (thejonroberts) [inline spec/views/notifications/index.html.erb_spec.rb:4]: what's going on here? seems unrelated?

File: spec/views/placements/edit.html.erb_spec.rb
Comment (thejonroberts) [inline spec/views/placements/edit.html.erb_spec.rb:19]:
```suggestion
    subject
```
?

tbh I haven't worked with view specs much, but I think render here is grabbing described_class by default, so use subject or use render, but subject is defined above.

**Review by thejonroberts — COMMENTED** [review]
Got a lot of comments, some are a little nit picky. But this is good overall! Will pull this down and try it out (probably tomorrow) before approval, but not requesting changes.

**Comment by guswhitten** [comment]
@thejonroberts i appreciate your effort, but it's functional and tested. this many refactorizations that change nothing about the actual product seems a bit ridiculous?

**Comment by thejonroberts** [comment]
> @thejonroberts i appreciate your effort, but it's functional and tested. this many refactorizations that change nothing about the actual product seems a bit ridiculous?

@guswhitten I hid some of the comments that were less important imo (I can't resolve convos). 

I do think that maintainability is important in addition to functionality - whether things are readable and modifiable in the future matters. That's the motivation behind most of the suggestions.  But I take your point. I didn't block merge by requesting changes. The suggestions are my opinions, the decision to use them is yours. I will approve when I can pull down and try it out, regardless.

I also made formatted suggestions so that you could commit directly from GitHub if you agree with the change, requiring less effort for you. Try it out with the two `placement_types`/`placement_names` changes in the factory if you like. I think that is definitely confusing to a reader because `placement_type` is a model/factory, and therefore prone to breaking something if changed/used incorrectly.

**Review by elasticspoon — COMMENTED** [review]
@guswhitten Hey sorry bout the delay reviewing this. Largely I am fine with the whole thing. It works nice in local testing.

I have 1 remaining concern which is mainly around not confusing users. Because the implementation order is a bit wack and orgs don't actually currently have a way to create placement types could you make this bit:

image
not show up. Either via flipper flag or just checking if an org has placements.

I'll leave it up to you on how much of the refactorings you want to accept but none of that stuff is blocking.

File: spec/system/placements/edit_spec.rb
Comment (elasticspoon) [reply spec/system/placements/edit_spec.rb:17]: I would argue that those are two different tests. Navigating to the index first catches any issues with the edit link not being present etc.

That being said that might already be covered in another test and be unnecessary.

**Review by elasticspoon — APPROVED** [review]
Actually after thinking a bit my earlier comment still applies but I would rather have this in sooner rather than later even in a slightly imperfect form. 

Feel free to add the conditional on the org having placement types and any other refactorings, otherwise I'll probably merge it in tomorrow and make an issue to not show the html until a organization has placements.

File: spec/system/placements/edit_spec.rb
Comment (thejonroberts) [reply spec/system/placements/edit_spec.rb:17]: yes, but I would say that is a concern of the index system or view spec... `it "displays link to edit" do` or similar there... edit spec shouldn't worry about index (or the other ways we can get to this page) at all imo.

File: app/views/casa_cases/_placements.html.erb
Comment (guswhitten) [reply app/views/casa_cases/_placements.html.erb:6]: to keep it similar to other casa_case partials. the `h6` includes Current Placement and the placement_type, but now excludes the link to "See All Placements" (though its CSS was being overridden anyway)

File: spec/views/notifications/index.html.erb_spec.rb
Comment (guswhitten) [reply spec/views/notifications/index.html.erb_spec.rb:4]: yeah this is unrelated. fixes a flaky test (from a previous PR of mine)

**Review by thejonroberts — APPROVED** [review]
Thank you for the scope and factory changes! Looks good!

**Comment by elasticspoon** [comment]
:+1: thank you for the work @guswhitten and @thejonroberts for the in depth review.

**Pull request merged**
